### PR TITLE
Remove 'get node' call during bootstrapping

### PR DESCRIPTION
**Pull request by liggitt**
Kubelet bootstrapping should have minimal permissions until it obtains an approved client certificate.

@luxas PTAL

/cc @mikedanese @cjcullen https://github.com/kubernetes/kubernetes/pull/40760#issuecomment-276832957

### CODE REVIEW

**Review by luxas — CHANGES_REQUESTED** [review]
Please don't remove this functionality. We need this kind of check, but if it's a must to remove the check from before it fetches the csr, at least do it after the tls bootstrap.

Still, I understand the reasoning behind it, but I'm against removing this functionality.

cc @mikedanese

**Comment by liggitt** [comment]
> Please don't remove this functionality. We need this kind of check, but if it's a must to remove the check from before it fetches the csr, at least do it after the tls bootstrap.

The current check is actually allowing existing nodes with identical names (if it successfully looks up the node object, `checkForNodeNameDuplicates` returns nil).

Even if a node object with this name already exists, this check would prevent re-bootstrapping the kubelet on the same machine. That seems wrong to me

**Comment by mikedanese** [comment]
/approve

> Even if a node object with this name already exists, this check would prevent re-bootstrapping the kubelet on the same machine. That seems wrong to me

I agree with this.

**Comment by cjcullen** [comment]
I think I agree w/ this PR, but I'd like to get agreement from @luxas.

My view:
When using a CSR group auto-approver, the csr.create permission is a pretty privileged thing, so including nodes.get isn't all that bad (csr.create can be trivially escalated into nodes.get).

In the long term, we will have a more thorough CSR approval process, so csr.create can be allowed to everyone (barring some protections for DOS, but that doesn't have to be in authz). When that happens, we don't want anonymous users to also have nodes.get. Maybe we do this by limiting the scope of the node-bootstrapping role now, and make kube-adm do without nodes.get. Maybe we do this with another built-in role for anonymous node-joining later, and allow kube-adm to keep using the existing node-bootstrapper role.

**Comment by liggitt** [comment]
> When using a CSR group auto-approver, the csr.create permission is a pretty privileged thing, so including nodes.get isn't all that bad (csr.create can be trivially escalated into nodes.get).
> 
> In the long term, we will have a more thorough CSR approval process

Right, the `system:node-bootstrapper` role should not presume auto-approve.

kubeadm could check for the node api object using the client credential obtained during bootstrap if it really wanted to, but I don't even think that is right (since it prevents re-bootstrapping a node). The kubelet accomodates its Node API object already existing, I think bootstrap should not add more restrictions on top of that.

**Comment by liggitt** [comment]
>> Even if a node object with this name already exists, this check would prevent re-bootstrapping the kubelet on the same machine. That seems wrong to me
>
> I agree with this.

@luxas given that, are you ok with this change?

**Review by luxas — APPROVED** [review]
@liggitt Ok, let's merge this then. I'm still not totally happy with it, but I guess it's the best we can do actually.

The reason we have (or had, I see the code isn't working correctly currently), is that, yes, there are users who are trying to join multiple nodes _with the same name_ and we've got "bug reports" for that.
Anyway, it might be worth it to try to solve it on an other layer in the k8s ecosystem than kubeadm...

**Comment by liggitt** [comment]
flake https://github.com/kubernetes/kubernetes/issues/37982
@k8s-bot bazel test this

**Comment by luxas** [comment]
@k8s-bot unit test this

**Pull request merged**
